File: server/FunctionalTests/Tests/Rojo05/LocalPull.py
"""
TheNexusAvenger

Tests that the local pull request acts correctly.
"""

import json
import NexusGitFunctionalTest



"""
Tests the local pull as a standalone request with several files.
"""
class Rojo05TestLocalPullAsStandaloneRequestMultipleFile(NexusGitFunctionalTest.NexusGitFunctionalTest):
    """
    Setup for the test.
    """
    def setupTest(self):
        self.workspace.writeFile("default.project.json","{\"name\": \"Repository\",\"servePort\": 20001,\"tree\": {\"$className\": \"DataModel\",\"ReplicatedStorage\": {\"$className\": \"ReplicatedStorage\",\"NexusGit\": {\"$path\": \"src\"},\"NexusGitTests\": {\"$path\": \"test\"}}}}")
        self.workspace.createDirectory("src")
        self.workspace.createDirectory("test")
        self.workspace.writeFile("src/testscript1.server.lua","print(\"Hello world 1!\")")
        self.workspace.writeFile("src/testscript2.client.lua","print(\"Hello world 2!\")")
        self.workspace.writeFile("test/testscript3.lua","print(\"Hello world 3!\")")

    """
    Runs the test.
    """
    def runTest(self):
        # Wait for it to initialize.
        self.setPortNumber(20001)
        self.waitForInitialization()

        # Send a standalone request and get the response.
        responseJson = self.sendGETRequest("/localpull")
        response = json.loads(responseJson)

        self.assertEqual(response["Type"],"Partitions")
        self.assertEqual(len(response["Instances"]),1)
        self.assertEqual(response["Instances"]["ReplicatedStorage"]["Properties"]["Name"]["Type"],"String")
        self.assertEqual(response["Instances"]["ReplicatedStorage"]["Properties"]["Name"]["Value"],"ReplicatedStorage")
        self.assertEqual(len(response["Instances"]["ReplicatedStorage"]["Children"]),2)

        self.assertEqual(response["Instances"]["ReplicatedStorage"]["Children"][0]["Properties"]["Name"]["Type"],"String")
        self.assertEqual(response["Instances"]["ReplicatedStorage"]["Children"][0]["Properties"]["Name"]["Value"],"NexusGit")
        self.assertEqual(len(response["Instances"]["ReplicatedStorage"]["Children"][0]["Children"]),2)

        self.assertEqual(response["Instances"]["ReplicatedStorage"]["Children"][1]["Properties"]["Name"]["Type"],"String")
        self.assertEqual(response["Instances"]["ReplicatedStorage"]["Children"][1]["Properties"]["Name"]["Value"],"NexusGitTests")
        self.assertEqual(len(response["Instances"]["ReplicatedStorage"]["Children"][1]["Children"]),1)

        self.assertEqual(response["Instances"]["ReplicatedStorage"]["Children"][0]["Children"][0]["Properties"]["ClassName"]["Type"],"String")
        self.assertEqual(response["Instances"]["ReplicatedStorage"]["Children"][0]["Children"][0]["Properties"]["ClassName"]["Value"],"Script")
        self.assertEqual(response["Instances"]["ReplicatedStorage"]["Children"][0]["Children"][0]["Properties"]["Name"]["Type"],"String")
        self.assertEqual(response["Instances"]["ReplicatedStorage"]["Children"][0]["Children"][0]["Properties"]["Name"]["Value"],"testscript1")
        self.assertEqual(response["Instances"]["ReplicatedStorage"]["Children"][0]["Children"][0]["Properties"]["Source"]["Type"],"String")
        self.assertEqual(response["Instances"]["ReplicatedStorage"]["Children"][0]["Children"][0]["Properties"]["Source"]["Value"],"print(\"Hello world 1!\")")
        self.assertEqual(len(response["Instances"]["ReplicatedStorage"]["Children"][0]["Children"][0]["Children"]),0)

        self.assertEqual(response["Instances"]["ReplicatedStorage"]["Children"][0]["Children"][1]["Properties"]["ClassName"]["Type"],"String")
        self.assertEqual(response["Instances"]["ReplicatedStorage"]["Children"][0]["Children"][1]["Properties"]["ClassName"]["Value"],"LocalScript")
        self.assertEqual(response["Instances"]["ReplicatedStorage"]["Children"][0]["Children"][1]["Properties"]["Name"]["Type"],"String")
        self.assertEqual(response["Instances"]["ReplicatedStorage"]["Children"][0]["Children"][1]["Properties"]["Name"]["Value"],"testscript2")
        self.assertEqual(response["Instances"]["ReplicatedStorage"]["Children"][0]["Children"][1]["Properties"]["Source"]["Type"],"String")
        self.assertEqual(response["Instances"]["ReplicatedStorage"]["Children"][0]["Children"][1]["Properties"]["Source"]["Value"],"print(\"Hello world 2!\")")
        self.assertEqual(len(response["Instances"]["ReplicatedStorage"]["Children"][0]["Children"][1]["Children"]),0)

        self.assertEqual(response["Instances"]["ReplicatedStorage"]["Children"][1]["Children"][0]["Properties"]["ClassName"]["Type"],"String")
        self.assertEqual(response["Instances"]["ReplicatedStorage"]["Children"][1]["Children"][0]["Properties"]["ClassName"]["Value"],"ModuleScript")
        self.assertEqual(response["Instances"]["ReplicatedStorage"]["Children"][1]["Children"][0]["Properties"]["Name"]["Type"],"String")
        self.assertEqual(response["Instances"]["ReplicatedStorage"]["Children"][1]["Children"][0]["Properties"]["Name"]["Value"],"testscript3")
        self.assertEqual(response["Instances"]["ReplicatedStorage"]["Children"][1]["Children"][0]["Properties"]["Source"]["Type"],"String")
        self.assertEqual(response["Instances"]["ReplicatedStorage"]["Children"][1]["Children"][0]["Properties"]["Source"]["Value"],"print(\"Hello world 3!\")")
        self.assertEqual(len(response["Instances"]["ReplicatedStorage"]["Children"][1]["Children"][0]["Children"]),0)

"""
Tests the local pull as a standalone request with directories.
"""
class Rojo05TestLocalPullAsStandaloneRequestDirectories(NexusGitFunctionalTest.NexusGitFunctionalTest):
    """
    Setup for the test.
    """
    def setupTest(self):
        self.workspace.writeFile("default.project.json","{\"name\": \"Repository\",\"servePort\": 20001,\"tree\": {\"$className\": \"DataModel\",\"ReplicatedStorage\": {\"$className\": \"ReplicatedStorage\",\"NexusGit\": {\"$path\": \"src\"},\"NexusGitTests\": {\"$path\": \"test\"}}}}")
        self.workspace.createDirectory("src")
        self.workspace.createDirectory("src/directory1")
        self.workspace.createDirectory("src/directory2")
        self.workspace.writeFile("src/directory1/testscript1.server.lua","print(\"Hello world 1!\")")
        self.workspace.writeFile("src/directory2/init.client.lua","print(\"Hello world 2!\")")
        self.workspace.writeFile("src/directory2/testscript3.lua","print(\"Hello world 3!\")")

    """
    Runs the test.
    """
    def runTest(self):
        # Wait for it to initialize.
        self.setPortNumber(20001)
        self.waitForInitialization()

        # Send a standalone request and get the response.
        responseJson = self.sendGETRequest("/localpull")
        response = json.loads(responseJson)

        self.assertEqual(response["Type"],"Partitions")
        self.assertEqual(len(response["Instances"]),1)
        self.assertEqual(response["Instances"]["ReplicatedStorage"]["Properties"]["Name"]["Type"],"String")
        self.assertEqual(response["Instances"]["ReplicatedStorage"]["Properties"]["Name"]["Value"],"ReplicatedStorage")
        self.assertEqual(len(response["Instances"]["ReplicatedStorage"]["Children"]),1)

        self.assertEqual(response["Instances"]["ReplicatedStorage"]["Children"][0]["Properties"]["ClassName"]["Type"],"String")
        self.assertEqual(response["Instances"]["ReplicatedStorage"]["Children"][0]["Properties"]["ClassName"]["Value"],"Folder")
        self.assertEqual(response["Instances"]["ReplicatedStorage"]["Children"][0]["Properties"]["Name"]["Type"],"String")
        self.assertEqual(response["Instances"]["ReplicatedStorage"]["Children"][0]["Properties"]["Name"]["Value"],"NexusGit")
        self.assertEqual(len(response["Instances"]["ReplicatedStorage"]["Children"][0]["Children"]),2)

        self.assertEqual(response["Instances"]["ReplicatedStorage"]["Children"][0]["Children"][0]["Properties"]["ClassName"]["Type"],"String")
        self.assertEqual(response["Instances"]["ReplicatedStorage"]["Children"][0]["Children"][0]["Properties"]["ClassName"]["Value"],"Folder")
        self.assertEqual(response["Instances"]["ReplicatedStorage"]["Children"][0]["Children"][0]["Properties"]["Name"]["Type"],"String")
        self.assertEqual(response["Instances"]["ReplicatedStorage"]["Children"][0]["Children"][0]["Properties"]["Name"]["Value"],"directory1")
        self.assertEqual(len(response["Instances"]["ReplicatedStorage"]["Children"][0]["Children"][0]["Children"]),1)

        self.assertEqual(response["Instances"]["ReplicatedStorage"]["Children"][0]["Children"][0]["Children"][0]["Properties"]["ClassName"]["Type"],"String")
        self.assertEqual(response["Instances"]["ReplicatedStorage"]["Children"][0]["Children"][0]["Children"][0]["Properties"]["ClassName"]["Value"],"Script")
        self.assertEqual(response["Instances"]["ReplicatedStorage"]["Children"][0]["Children"][0]["Children"][0]["Properties"]["Name"]["Type"],"String")
        self.assertEqual(response["Instances"]["ReplicatedStorage"]["Children"][0]["Children"][0]["Children"][0]["Properties"]["Name"]["Value"],"testscript1")
        self.assertEqual(response["Instances"]["ReplicatedStorage"]["Children"][0]["Children"][0]["Children"][0]["Properties"]["Source"]["Type"],"String")
        self.assertEqual(response["Instances"]["ReplicatedStorage"]["Children"][0]["Children"][0]["Children"][0]["Properties"]["Source"]["Value"],"print(\"Hello world 1!\")")
        self.assertEqual(len(response["Instances"]["ReplicatedStorage"]["Children"][0]["Children"][0]["Children"][0]["Children"]),0)

        self.assertEqual(response["Instances"]["ReplicatedStorage"]["Children"][0]["Children"][1]["Properties"]["ClassName"]["Type"],"String")
        self.assertEqual(response["Instances"]["ReplicatedStorage"]["Children"][0]["Children"][1]["Properties"]["ClassName"]["Value"],"LocalScript")
        self.assertEqual(response["Instances"]["ReplicatedStorage"]["Children"][0]["Children"][1]["Properties"]["Name"]["Type"],"String")
        self.assertEqual(response["Instances"]["ReplicatedStorage"]["Children"][0]["Children"][1]["Properties"]["Name"]["Value"],"directory2")
        self.assertEqual(response["Instances"]["ReplicatedStorage"]["Children"][0]["Children"][1]["Properties"]["Source"]["Type"],"String")
        self.assertEqual(response["Instances"]["ReplicatedStorage"]["Children"][0]["Children"][1]["Properties"]["Source"]["Value"],"print(\"Hello world 2!\")")
        self.assertEqual(len(response["Instances"]["ReplicatedStorage"]["Children"][0]["Children"][1]["Children"]),1)

        self.assertEqual(response["Instances"]["ReplicatedStorage"]["Children"][0]["Children"][1]["Children"][0]["Properties"]["ClassName"]["Type"],"String")
        self.assertEqual(response["Instances"]["ReplicatedStorage"]["Children"][0]["Children"][1]["Children"][0]["Properties"]["ClassName"]["Value"],"ModuleScript")
        self.assertEqual(response["Instances"]["ReplicatedStorage"]["Children"][0]["Children"][1]["Children"][0]["Properties"]["Name"]["Type"],"String")
        self.assertEqual(response["Instances"]["ReplicatedStorage"]["Children"][0]["Children"][1]["Children"][0]["Properties"]["Name"]["Value"],"testscript3")
        self.assertEqual(response["Instances"]["ReplicatedStorage"]["Children"][0]["Children"][1]["Children"][0]["Properties"]["Source"]["Type"],"String")
        self.assertEqual(response["Instances"]["ReplicatedStorage"]["Children"][0]["Children"][1]["Children"][0]["Properties"]["Source"]["Value"],"print(\"Hello world 3!\")")
        self.assertEqual(len(response["Instances"]["ReplicatedStorage"]["Children"][0]["Children"][1]["Children"][0]["Children"]),0)

"""
Tests the local pull as a standalone request with a JSON model file.
"""
class Rojo05TestLocalPullAsStandaloneRequestJSONModelFile(NexusGitFunctionalTest.NexusGitFunctionalTest):
    """
    Setup for the test.
    """
    def setupTest(self):
        self.workspace.writeFile("default.project.json","{\"name\": \"Repository\",\"servePort\": 20001,\"tree\": {\"$className\": \"DataModel\",\"ReplicatedStorage\": {\"$className\": \"ReplicatedStorage\",\"NexusGit\": {\"$path\": \"src\"},\"NexusGitTests\": {\"$path\": \"test\"}}}}")
        self.workspace.createDirectory("src")
        self.workspace.createDirectory("src/directory")
        self.workspace.writeFile("src/directory/init.server.lua","print(\"Hello world!\")")
        self.workspace.writeFile("src/directory/hello.model.json","{\"Name\": \"hello\",\"ClassName\": \"Model\",\"Children\": [{\"Name\": \"Some Part\",\"ClassName\": \"Part\"},{\"Name\": \"Some StringValue\",\"ClassName\": \"StringValue\",\"Properties\": {\"Value\": {\"Type\": \"String\",\"Value\": \"Hello, world!\"}}}]}")

    """
    Runs the test.
    """
    def runTest(self):
        # Wait for it to initialize.
        self.setPortNumber(20001)
        self.waitForInitialization()

        # Send a standalone request and get the response.
        responseJson = self.sendGETRequest("/localpull")
        response = json.loads(responseJson)

        self.assertEqual(response["Type"],"Partitions")
        self.assertEqual(len(response["Instances"]),1)
        self.assertEqual(response["Instances"]["ReplicatedStorage"]["Properties"]["Name"]["Type"],"String")
        self.assertEqual(response["Instances"]["ReplicatedStorage"]["Properties"]["Name"]["Value"],"ReplicatedStorage")
        self.assertEqual(len(response["Instances"]["ReplicatedStorage"]["Children"]),1)

        self.assertEqual(response["Instances"]["ReplicatedStorage"]["Children"][0]["Properties"]["Name"]["Type"],"String")
        self.assertEqual(response["Instances"]["ReplicatedStorage"]["Children"][0]["Properties"]["Name"]["Value"],"NexusGit")
        self.assertEqual(len(response["Instances"]["ReplicatedStorage"]["Children"][0]["Children"]),1)

        self.assertEqual(response["Instances"]["ReplicatedStorage"]["Children"][0]["Children"][0]["Properties"]["ClassName"]["Type"],"String")
        self.assertEqual(response["Instances"]["ReplicatedStorage"]["Children"][0]["Children"][0]["Properties"]["ClassName"]["Value"],"Script")
        self.assertEqual(response["Instances"]["ReplicatedStorage"]["Children"][0]["Children"][0]["Properties"]["Name"]["Type"],"String")
        self.assertEqual(response["Instances"]["ReplicatedStorage"]["Children"][0]["Children"][0]["Properties"]["Name"]["Value"],"directory")
        self.assertEqual(response["Instances"]["ReplicatedStorage"]["Children"][0]["Children"][0]["Properties"]["Source"]["Type"],"String")
        self.assertEqual(response["Instances"]["ReplicatedStorage"]["Children"][0]["Children"][0]["Properties"]["Source"]["Value"],"print(\"Hello world!\")")
        self.assertEqual(len(response["Instances"]["ReplicatedStorage"]["Children"][0]["Children"][0]["Children"]),1)

        self.assertEqual(response["Instances"]["ReplicatedStorage"]["Children"][0]["Children"][0]["Children"][0]["Properties"]["ClassName"]["Type"],"String")
        self.assertEqual(response["Instances"]["ReplicatedStorage"]["Children"][0]["Children"][0]["Children"][0]["Properties"]["ClassName"]["Value"],"Model")
        self.assertEqual(response["Instances"]["ReplicatedStorage"]["Children"][0]["Children"][0]["Children"][0]["Properties"]["Name"]["Type"],"String")
        self.assertEqual(response["Instances"]["ReplicatedStorage"]["Children"][0]["Children"][0]["Children"][0]["Properties"]["Name"]["Value"],"hello")
        self.assertEqual(len(response["Instances"]["ReplicatedStorage"]["Children"][0]["Children"][0]["Children"][0]["Children"]),2)

        self.assertEqual(response["Instances"]["ReplicatedStorage"]["Children"][0]["Children"][0]["Children"][0]["Children"][0]["Properties"]["ClassName"]["Type"],"String")
        self.assertEqual(response["Instances"]["ReplicatedStorage"]["Children"][0]["Children"][0]["Children"][0]["Children"][0]["Properties"]["ClassName"]["Value"],"Part")
        self.assertEqual(response["Instances"]["ReplicatedStorage"]["Children"][0]["Children"][0]["Children"][0]["Children"][0]["Properties"]["Name"]["Type"],"String")
        self.assertEqual(response["Instances"]["ReplicatedStorage"]["Children"][0]["Children"][0]["Children"][0]["Children"][0]["Properties"]["Name"]["Value"],"Some Part")
        self.assertEqual(len(response["Instances"]["ReplicatedStorage"]["Children"][0]["Children"][0]["Children"][0]["Children"][0]["Children"]),0)

        self.assertEqual(response["Instances"]["ReplicatedStorage"]["Children"][0]["Children"][0]["Children"][0]["Children"][1]["Properties"]["ClassName"]["Type"],"String")
        self.assertEqual(response["Instances"]["ReplicatedStorage"]["Children"][0]["Children"][0]["Children"][0]["Children"][1]["Properties"]["ClassName"]["Value"],"StringValue")
        self.assertEqual(response["Instances"]["ReplicatedStorage"]["Children"][0]["Children"][0]["Children"][0]["Children"][1]["Properties"]["Name"]["Type"],"String")
        self.assertEqual(response["Instances"]["ReplicatedStorage"]["Children"][0]["Children"][0]["Children"][0]["Children"][1]["Properties"]["Name"]["Value"],"Some StringValue")
        self.assertEqual(response["Instances"]["ReplicatedStorage"]["Children"][0]["Children"][0]["Children"][0]["Children"][1]["Properties"]["Value"]["Type"],"String")
        self.assertEqual(response["Instances"]["ReplicatedStorage"]["Children"][0]["Children"][0]["Children"][0]["Children"][1]["Properties"]["Value"]["Value"],"Hello, world!")
        self.assertEqual(len(response["Instances"]["ReplicatedStorage"]["Children"][0]["Children"][0]["Children"][0]["Children"][1]["Children"]),0)

"""
Tests the local pull as a standalone request with a text file.
"""
class Rojo05TestLocalPullAsStandaloneRequestTxtFile(NexusGitFunctionalTest.NexusGitFunctionalTest):
    """
    Setup for the test.
    """
    def setupTest(self):
        self.workspace.writeFile("default.project.json","{\"name\": \"Repository\",\"servePort\": 20001,\"tree\": {\"$className\": \"DataModel\",\"ReplicatedStorage\": {\"$className\": \"ReplicatedStorage\",\"NexusGit\": {\"$path\": \"src\"},\"NexusGitTests\": {\"$path\": \"test\"}}}}")
        self.workspace.createDirectory("src")
        self.workspace.writeFile("src/TestValue.txt","Test value")

    """
    Runs the test.
    """
    def runTest(self):
        # Wait for it to initialize.
        self.setPortNumber(20001)
        self.waitForInitialization()

        # Send a standalone request and get the response.
        responseJson = self.sendGETRequest("/localpull")
        response = json.loads(responseJson)

        self.assertEqual(response["Type"],"Partitions")
        self.assertEqual(len(response["Instances"]),1)
        self.assertEqual(response["Instances"]["ReplicatedStorage"]["Properties"]["Name"]["Type"],"String")
        self.assertEqual(response["Instances"]["ReplicatedStorage"]["Properties"]["Name"]["Value"],"ReplicatedStorage")
        self.assertEqual(len(response["Instances"]["ReplicatedStorage"]["Children"]),1)

        self.assertEqual(response["Instances"]["ReplicatedStorage"]["Children"][0]["Properties"]["ClassName"]["Type"],"String")
        self.assertEqual(response["Instances"]["ReplicatedStorage"]["Children"][0]["Properties"]["ClassName"]["Value"],"Folder")
        self.assertEqual(response["Instances"]["ReplicatedStorage"]["Children"][0]["Properties"]["Name"]["Type"],"String")
        self.assertEqual(response["Instances"]["ReplicatedStorage"]["Children"][0]["Properties"]["Name"]["Value"],"NexusGit")
        self.assertEqual(len(response["Instances"]["ReplicatedStorage"]["Children"][0]["Children"]),1)

        self.assertEqual(response["Instances"]["ReplicatedStorage"]["Children"][0]["Children"][0]["Properties"]["ClassName"]["Type"],"String")
        self.assertEqual(response["Instances"]["ReplicatedStorage"]["Children"][0]["Children"][0]["Properties"]["ClassName"]["Value"], "StringValue")
        self.assertEqual(response["Instances"]["ReplicatedStorage"]["Children"][0]["Children"][0]["Properties"]["Name"]["Type"],"String")
        self.assertEqual(response["Instances"]["ReplicatedStorage"]["Children"][0]["Children"][0]["Properties"]["Name"]["Value"],"TestValue")
        self.assertEqual(response["Instances"]["ReplicatedStorage"]["Children"][0]["Children"][0]["Properties"]["Value"]["Type"],"String")
        self.assertEqual(response["Instances"]["ReplicatedStorage"]["Children"][0]["Children"][0]["Properties"]["Value"]["Value"],"Test value")
        self.assertEqual(len(response["Instances"]["ReplicatedStorage"]["Children"][0]["Children"][0]["Children"]),0)

File: server/NexusGitTests/Functional/Rojo05/LocalPull.cs
/*
 * TheNexusAvenger
 *
 * Tests the LocalPull request with Rojo 0.5.
 */

using Newtonsoft.Json;
using NexusGit.RobloxInstance;
using NUnit.Framework;

namespace NexusGitTests.Functional.Rojo05
{
    /*
     * Tests local pull as a standalone request.
     */
    public class TestLocalPull : BaseFunctionalTest
    {
        /*
         * Sets up the test.
         */
        public override void Setup()
        {
            // Write the files.
            this.Workspace.WriteFile("default.project.json","{\"name\": \"Repository\",\"servePort\": 20001,\"tree\": {\"$className\": \"DataModel\",\"ReplicatedStorage\": {\"$className\": \"ReplicatedStorage\",\"NexusGit\": {\"$path\": \"src\"},\"NexusGitTests\": {\"$path\": \"test\"}}}}");
            this.Workspace.CreateDirectory("src");
            this.Workspace.CreateDirectory("test");
            this.Workspace.WriteFile("src/testscript1.server.lua","print(\"Hello world 1!\")");
            this.Workspace.WriteFile("src/testscript2.client.lua", "print(\"Hello world 2!\")");
            this.Workspace.WriteFile("test/testscript3.lua", "print(\"Hello world 3!\")");
        }
        
        /*
         * Runs the test.
         */
        public override void Run()
        {
            var response = JsonConvert.DeserializeObject<Partitions>(this.SendGETRequest("/localpull"));

            Assert.AreEqual(response.Type,"Partitions");
            Assert.AreEqual(response.Instances.Count,2);
            Assert.AreEqual(response.Instances["ReplicatedStorage.NexusGit"].Properties["Name"].Type,"String");
            Assert.AreEqual(response.Instances["ReplicatedStorage.NexusGit"].Properties["Name"].Value,"NexusGit");
            Assert.AreEqual(response.Instances["ReplicatedStorage.NexusGit"].Children.Count,2);
            Assert.AreEqual(response.Instances["ReplicatedStorage.NexusGitTest"].Children.Count,1);

            Assert.AreEqual(response.Instances["ReplicatedStorage.NexusGit"].Children[0].Properties["ClassName"].Type,"String");
            Assert.AreEqual(response.Instances["ReplicatedStorage.NexusGit"].Children[0].Properties["ClassName"].Value,"Script");
            Assert.AreEqual(response.Instances["ReplicatedStorage.NexusGit"].Children[0].Properties["Name"].Type,"String");
            Assert.AreEqual(response.Instances["ReplicatedStorage.NexusGit"].Children[0].Properties["Name"].Value,"testscript1");
            Assert.AreEqual(response.Instances["ReplicatedStorage.NexusGit"].Children[0].Properties["Source"].Type,"String");
            Assert.AreEqual(response.Instances["ReplicatedStorage.NexusGit"].Children[0].Properties["Source"].Value,"print(\"Hello world 1!\")");
            Assert.AreEqual(response.Instances["ReplicatedStorage.NexusGit"].Children[0].Children.Count,0);

            Assert.AreEqual(response.Instances["ReplicatedStorage.NexusGit"].Children[1].Properties["ClassName"].Type,"String");
            Assert.AreEqual(response.Instances["ReplicatedStorage.NexusGit"].Children[1].Properties["ClassName"].Value,"LocalScript");
            Assert.AreEqual(response.Instances["ReplicatedStorage.NexusGit"].Children[1].Properties["Name"].Type,"String");
            Assert.AreEqual(response.Instances["ReplicatedStorage.NexusGit"].Children[1].Properties["Name"].Value,"testscript2");
            Assert.AreEqual(response.Instances["ReplicatedStorage.NexusGit"].Children[1].Properties["Source"].Type,"String");
            Assert.AreEqual(response.Instances["ReplicatedStorage.NexusGit"].Children[1].Properties["Source"].Value,"print(\"Hello world 2!\")");
            Assert.AreEqual(response.Instances["ReplicatedStorage.NexusGit"].Children[1].Children.Count,0);

            Assert.AreEqual(response.Instances["ReplicatedStorage.NexusGitTest"].Children[0].Properties["ClassName"].Type,"String");
            Assert.AreEqual(response.Instances["ReplicatedStorage.NexusGitTest"].Children[0].Properties["ClassName"].Value,"ModuleScript");
            Assert.AreEqual(response.Instances["ReplicatedStorage.NexusGitTest"].Children[0].Properties["Name"].Type,"String");
            Assert.AreEqual(response.Instances["ReplicatedStorage.NexusGitTest"].Children[0].Properties["Name"].Value,"testscript3");
            Assert.AreEqual(response.Instances["ReplicatedStorage.NexusGitTest"].Children[0].Properties["Source"].Type,"String");
            Assert.AreEqual(response.Instances["ReplicatedStorage.NexusGitTest"].Children[0].Properties["Source"].Value,"print(\"Hello world 3!\")");
            Assert.AreEqual(response.Instances["ReplicatedStorage.NexusGitTest"].Children[0].Children.Count,0);
        }
    }
    
    /*
     * Tests local pull with directories as a standalone request.
     */
    public class TestLocalPullWithDirectories : BaseFunctionalTest
    {
        /*
         * Sets up the test.
         */
        public override void Setup()
        {
            // Write the files.
            this.Workspace.WriteFile("default.project.json","{\"name\": \"Repository\",\"servePort\": 20001,\"tree\": {\"$className\": \"DataModel\",\"ReplicatedStorage\": {\"$className\": \"ReplicatedStorage\",\"NexusGit\": {\"$path\": \"src\"},\"NexusGitTests\": {\"$path\": \"test\"}}}}");
            this.Workspace.CreateDirectory("src");
            this.Workspace.CreateDirectory("src/directory1");
            this.Workspace.CreateDirectory("src/directory2");
            this.Workspace.WriteFile("src/directory1/testscript1.server.lua","print(\"Hello world 1!\")");
            this.Workspace.WriteFile("src/directory1/testscript2.client.lua", "print(\"Hello world 2!\")");
            this.Workspace.WriteFile("src/directory2/testscript3.lua", "print(\"Hello world 3!\")");
        }
        
        /*
         * Runs the test.
         */
        public override void Run()
        {
            var response = JsonConvert.DeserializeObject<Partitions>(this.SendGETRequest("/localpull"));

            Assert.AreEqual(response.Type,"Partitions");
            Assert.AreEqual(response.Instances.Count,1);
            Assert.AreEqual(response.Instances["ReplicatedStorage.NexusGit"].Properties["Name"].Type,"String");
            Assert.AreEqual(response.Instances["ReplicatedStorage.NexusGit"].Properties["Name"].Value,"NexusGit");
            Assert.AreEqual(response.Instances["ReplicatedStorage.NexusGit"].Children,2);

            Assert.AreEqual(response.Instances["ReplicatedStorage.NexusGit"].Children[0].Properties["ClassName"].Type,"String");
            Assert.AreEqual(response.Instances["ReplicatedStorage.NexusGit"].Children[0].Properties["ClassName"].Value,"Folder");
            Assert.AreEqual(response.Instances["ReplicatedStorage.NexusGit"].Children[0].Properties["Name"].Type,"String");
            Assert.AreEqual(response.Instances["ReplicatedStorage.NexusGit"].Children[0].Properties["Name"].Value,"directory1");
            Assert.AreEqual(response.Instances["ReplicatedStorage.NexusGit"].Children[0].Children,1);

            Assert.AreEqual(response.Instances["ReplicatedStorage.NexusGit"].Children[0].Children[0].Properties["ClassName"].Type,"String");
            Assert.AreEqual(response.Instances["ReplicatedStorage.NexusGit"].Children[0].Children[0].Properties["ClassName"].Value,"Script");
            Assert.AreEqual(response.Instances["ReplicatedStorage.NexusGit"].Children[0].Children[0].Properties["Name"].Type,"String");
            Assert.AreEqual(response.Instances["ReplicatedStorage.NexusGit"].Children[0].Children[0].Properties["Name"].Value,"testscript1");
            Assert.AreEqual(response.Instances["ReplicatedStorage.NexusGit"].Children[0].Children[0].Properties["Source"].Type,"String");
            Assert.AreEqual(response.Instances["ReplicatedStorage.NexusGit"].Children[0].Children[0].Properties["Source"].Value,"print(\"Hello world 1!\")");
            Assert.AreEqual(response.Instances["ReplicatedStorage.NexusGit"].Children[0].Children[0].Children,0);

            Assert.AreEqual(response.Instances["ReplicatedStorage.NexusGit"].Children[1].Properties["ClassName"].Type,"String");
            Assert.AreEqual(response.Instances["ReplicatedStorage.NexusGit"].Children[1].Properties["ClassName"].Value,"LocalScript");
            Assert.AreEqual(response.Instances["ReplicatedStorage.NexusGit"].Children[1].Properties["Name"].Type,"String");
            Assert.AreEqual(response.Instances["ReplicatedStorage.NexusGit"].Children[1].Properties["Name"].Value,"directory2");
            Assert.AreEqual(response.Instances["ReplicatedStorage.NexusGit"].Children[1].Properties["Source"].Type,"String");
            Assert.AreEqual(response.Instances["ReplicatedStorage.NexusGit"].Children[1].Properties["Source"].Value,"print(\"Hello world 2!\")");
            Assert.AreEqual(response.Instances["ReplicatedStorage.NexusGit"].Children[1].Children,1);

            Assert.AreEqual(response.Instances["ReplicatedStorage.NexusGit"].Children[1].Children[0].Properties["ClassName"].Type,"String");
            Assert.AreEqual(response.Instances["ReplicatedStorage.NexusGit"].Children[1].Children[0].Properties["ClassName"].Value,"ModuleScript");
            Assert.AreEqual(response.Instances["ReplicatedStorage.NexusGit"].Children[1].Children[0].Properties["Name"].Type,"String");
            Assert.AreEqual(response.Instances["ReplicatedStorage.NexusGit"].Children[1].Children[0].Properties["Name"].Value,"testscript3");
            Assert.AreEqual(response.Instances["ReplicatedStorage.NexusGit"].Children[1].Children[0].Properties["Source"].Type,"String");
            Assert.AreEqual(response.Instances["ReplicatedStorage.NexusGit"].Children[1].Children[0].Properties["Source"].Value,"print(\"Hello world 3!\")");
            Assert.AreEqual(response.Instances["ReplicatedStorage.NexusGit"].Children[1].Children[0].Children,0);
        }
    }
    
    /*
     * Tests local pull with a model file as a standalone request.
     */
    public class TestLocalPullWithModelFile : BaseFunctionalTest
    {
        /*
         * Sets up the test.
         */
        public override void Setup()
        {
            // Write the files.
            this.Workspace.WriteFile("default.project.json","{\"name\": \"Repository\",\"servePort\": 20001,\"tree\": {\"$className\": \"DataModel\",\"ReplicatedStorage\": {\"$className\": \"ReplicatedStorage\",\"NexusGit\": {\"$path\": \"src\"},\"NexusGitTests\": {\"$path\": \"test\"}}}}");
            this.Workspace.CreateDirectory("src");
            this.Workspace.CreateDirectory("src/directory");
            this.Workspace.WriteFile("src/directory/init.server.lua","print(\"Hello world!\")");
            this.Workspace.WriteFile("src/directory/hello.model.json","{\"Name\": \"hello\",\"ClassName\": \"Model\",\"Children\": [{\"Name\": \"Some Part\",\"ClassName\": \"Part\"},{\"Name\": \"Some StringValue\",\"ClassName\": \"StringValue\",\"Properties\": {\"Value\": {\"Type\": \"String\",\"Value\": \"Hello, world!\"}}}]}");
        }
        
        /*
         * Runs the test.
         */
        public override void Run()
        {
            var response = JsonConvert.DeserializeObject<Partitions>(this.SendGETRequest("/localpull"));

            Assert.AreEqual(response.Type,"Partitions");
            Assert.AreEqual(response.Instances,1);
            Assert.AreEqual(response.Instances["ReplicatedStorage.NexusGit"].Properties["Name"].Type,"String");
            Assert.AreEqual(response.Instances["ReplicatedStorage.NexusGit"].Properties["Name"].Value,"NexusGit");
            Assert.AreEqual(response.Instances["ReplicatedStorage.NexusGit"].Children,1);

            Assert.AreEqual(response.Instances["ReplicatedStorage.NexusGit"].Children[0].Properties["ClassName"].Type,"String");
            Assert.AreEqual(response.Instances["ReplicatedStorage.NexusGit"].Children[0].Properties["ClassName"].Value,"Script");
            Assert.AreEqual(response.Instances["ReplicatedStorage.NexusGit"].Children[0].Properties["Name"].Type,"String");
            Assert.AreEqual(response.Instances["ReplicatedStorage.NexusGit"].Children[0].Properties["Name"].Value,"directory");
            Assert.AreEqual(response.Instances["ReplicatedStorage.NexusGit"].Children[0].Properties["Source"].Type,"String");
            Assert.AreEqual(response.Instances["ReplicatedStorage.NexusGit"].Children[0].Properties["Source"].Value,"print(\"Hello world!\")");
            Assert.AreEqual(response.Instances["ReplicatedStorage.NexusGit"].Children[0].Children,1);

            Assert.AreEqual(response.Instances["ReplicatedStorage.NexusGit"].Children[0].Children[0].Properties["ClassName"].Type,"String");
            Assert.AreEqual(response.Instances["ReplicatedStorage.NexusGit"].Children[0].Children[0].Properties["ClassName"].Value,"Model");
            Assert.AreEqual(response.Instances["ReplicatedStorage.NexusGit"].Children[0].Children[0].Properties["Name"].Type,"String");
            Assert.AreEqual(response.Instances["ReplicatedStorage.NexusGit"].Children[0].Children[0].Properties["Name"].Value,"hello");
            Assert.AreEqual(response.Instances["ReplicatedStorage.NexusGit"].Children[0].Children[0].Children,2);

            Assert.AreEqual(response.Instances["ReplicatedStorage.NexusGit"].Children[0].Children[0].Children[0].Properties["ClassName"].Type,"String");
            Assert.AreEqual(response.Instances["ReplicatedStorage.NexusGit"].Children[0].Children[0].Children[0].Properties["ClassName"].Value,"Part");
            Assert.AreEqual(response.Instances["ReplicatedStorage.NexusGit"].Children[0].Children[0].Children[0].Properties["Name"].Type,"String");
            Assert.AreEqual(response.Instances["ReplicatedStorage.NexusGit"].Children[0].Children[0].Children[0].Properties["Name"].Value,"Some Part");
            Assert.AreEqual(response.Instances["ReplicatedStorage.NexusGit"].Children[0].Children[0].Children[0].Children,0);

            Assert.AreEqual(response.Instances["ReplicatedStorage.NexusGit"].Children[0].Children[0].Children[1].Properties["ClassName"].Type,"String");
            Assert.AreEqual(response.Instances["ReplicatedStorage.NexusGit"].Children[0].Children[0].Children[1].Properties["ClassName"].Value,"StringValue");
            Assert.AreEqual(response.Instances["ReplicatedStorage.NexusGit"].Children[0].Children[0].Children[1].Properties["Name"].Type,"String");
            Assert.AreEqual(response.Instances["ReplicatedStorage.NexusGit"].Children[0].Children[0].Children[1].Properties["Name"].Value,"Some StringValue");
            Assert.AreEqual(response.Instances["ReplicatedStorage.NexusGit"].Children[0].Children[0].Children[1].Properties["Value"].Type,"String");
            Assert.AreEqual(response.Instances["ReplicatedStorage.NexusGit"].Children[0].Children[0].Children[1].Properties["Value"].Value,"Hello, world!");
            Assert.AreEqual(response.Instances["ReplicatedStorage.NexusGit"].Children[0].Children[0].Children[1].Children,0);
        }
    }
    
    /*
     * Tests local pull with a text file as a standalone request.
     */
    public class TestLocalPullWithTextFile : BaseFunctionalTest
    {
        /*
         * Sets up the test.
         */
        public override void Setup()
        {
            // Write the files.
            this.Workspace.WriteFile("default.project.json","{\"name\": \"Repository\",\"servePort\": 20001,\"tree\": {\"$className\": \"DataModel\",\"ReplicatedStorage\": {\"$className\": \"ReplicatedStorage\",\"NexusGit\": {\"$path\": \"src\"},\"NexusGitTests\": {\"$path\": \"test\"}}}}");
            this.Workspace.CreateDirectory("src");
            this.Workspace.WriteFile("src/TestValue.txt","Test value");
        }
        
        /*
         * Runs the test.
         */
        public override void Run()
        {
            var response = JsonConvert.DeserializeObject<Partitions>(this.SendGETRequest("/localpull"));

            Assert.AreEqual(response.Type,"Partitions");
            Assert.AreEqual(response.Instances,1);
            Assert.AreEqual(response.Instances["ReplicatedStorage.NexusGit"].Properties["Name"].Type,"String");
            Assert.AreEqual(response.Instances["ReplicatedStorage.NexusGit"].Properties["Name"].Value,"NexusGit");
            Assert.AreEqual(response.Instances["ReplicatedStorage.NexusGit"].Children,1);
            
            Assert.AreEqual(response.Instances["ReplicatedStorage"].Children[0].Properties["ClassName"].Type,"String");
            Assert.AreEqual(response.Instances["ReplicatedStorage"].Children[0].Properties["ClassName"].Value,"Folder");
            Assert.AreEqual(response.Instances["ReplicatedStorage"].Children[0].Properties["Name"].Type,"String");
            Assert.AreEqual(response.Instances["ReplicatedStorage"].Children[0].Properties["Name"].Value,"NexusGit");
            Assert.AreEqual(response.Instances["ReplicatedStorage"].Children[0].Children.Count,1);

            Assert.AreEqual(response.Instances["ReplicatedStorage"].Children[0].Children[0].Properties["ClassName"].Type,"String");
            Assert.AreEqual(response.Instances["ReplicatedStorage"].Children[0].Children[0].Properties["ClassName"].Value, "StringValue");
            Assert.AreEqual(response.Instances["ReplicatedStorage"].Children[0].Children[0].Properties["Name"].Type,"String");
            Assert.AreEqual(response.Instances["ReplicatedStorage"].Children[0].Children[0].Properties["Name"].Value,"TestValue");
            Assert.AreEqual(response.Instances["ReplicatedStorage"].Children[0].Children[0].Properties["Value"].Type,"String");
            Assert.AreEqual(response.Instances["ReplicatedStorage"].Children[0].Children[0].Properties["Value"].Value,"Test value");
            Assert.AreEqual(response.Instances["ReplicatedStorage"].Children[0].Children[0].Children.Count,0);
        }
    }
}